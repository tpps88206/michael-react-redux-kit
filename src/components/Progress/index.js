import React from 'react';

import classNames from 'classnames';

import CircularProgress from '@material-ui/core/CircularProgress';

const Progress = ({ message, className, style }) => (
  <div className={classNames('d-flex flex-column justify-center align-center vh-100', className)} style={style}>
    <CircularProgress />
    {message && (
      <>
        <br />
        <p>{message}</p>
      </>
    )}
  </div>
);

export default Progress;
