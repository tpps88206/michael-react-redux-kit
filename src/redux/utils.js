import { ofType } from 'redux-observable';

import flatten from 'lodash/flatten';
import fromPairs from 'lodash/fromPairs';
import isEmpty from 'lodash/isEmpty';
import { mergeMap } from 'rxjs/operators';

// there are two type of handler, different in key name:
// - `actionName`: map handler to `actionName`
// - `actionNameEpic`: get full control of epic function
const sliceToEpics = slice => {
  if (!slice || !slice.epics) {
    return [];
  }
  const actionTypes = fromPairs(
    Object.entries(slice)
      .filter(([k]) => k !== 'default' && k !== 'epics')
      .filter(([, v]) => v && v.type)
      .map(([k, v]) => [k, v.type]),
  );
  const epics = Object.entries(slice.epics).map(([k, handler]) => {
    if (k.endsWith('Epic')) {
      return handler;
    }

    return (action$, state$) => {
      return action$.pipe(
        ofType(actionTypes[k]),
        mergeMap(action => handler(action$, state$, action)),
      );
    };
  });

  return epics;
};

export const getEpicsFromSlices = slices => {
  if (isEmpty(slices)) {
    return [];
  }
  return flatten(slices.map(sliceToEpics));
};
