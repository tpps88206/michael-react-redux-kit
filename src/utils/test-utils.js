import React from 'react';
import { Provider } from 'react-redux';

import { create } from 'jss';
import preset from 'jss-preset-default';

import CssBaseline from '@material-ui/core/CssBaseline';
import { createMuiTheme, StylesProvider, ThemeProvider } from '@material-ui/core/styles';
import { render as rtlRender } from '@testing-library/react';

import theme from '@/constants/theme';
import configureStore from '@/redux/configureStore';
import styles from '@/styles';

const render = (ui, { ...renderOptions } = {}) => {
  const muiTheme = createMuiTheme(theme);
  const jss = create({
    plugins: [...preset().plugins],
  });
  const store = configureStore();

  jss.createStyleSheet(styles(muiTheme), { meta: 'global' }).attach();

  const Wrapper = ({ children }) => {
    return (
      <StylesProvider jss={jss}>
        <ThemeProvider theme={muiTheme}>
          <CssBaseline />
          <Provider store={store}>{children}</Provider>
        </ThemeProvider>
      </StylesProvider>
    );
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
