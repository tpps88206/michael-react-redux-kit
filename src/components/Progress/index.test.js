import React from 'react';

import { render } from '@/utils/test-utils';
import Progress from './index';

it('renders Progress without crashing', () => {
  const div = document.createElement('div');
  render(<Progress />, div);
});

it('renders Progress with snapshot', () => {
  const { asFragment } = render(<Progress />);
  expect(asFragment()).toMatchSnapshot();
});
