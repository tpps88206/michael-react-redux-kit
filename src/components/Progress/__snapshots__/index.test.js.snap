// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Progress with snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column justify-center align-center vh-100"
  >
    <div
      class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
